### DISCUSSION TITLE
Improve startup time and optimize cpu and memory requests and limits for kubernetes

### DISCUSSION THREAD

**Discussion by NilsEngelbach** [keycloak/keycloak#23231] [Q&A]
Hello,

We are currently running keycloak in our kubernetes cluster.

The image is build beforehand an started with `--otpimized` as described in the documentation. 
There is alaway a big cpu peak on startup and it takes quite long to be ready (couple minutes).

The limits are currently set like this:

```yaml
resources:
  requests:
    memory: "512Mi"
    cpu: "250m"
  limits:
    memory: "512Mi"
    cpu: "500m"
``` 

After the initial peak the service does not have a lot of load and would not require such extensive resource definitions.

Are there any guidelines how to configure the resources?
And are there any things that could be done to improve the startup time (e.g. disabling unused features)?

Thanks in advance!

**Comment by shawkins** [comment]
> There is alaway a big cpu peak on startup and it takes quite long to be ready (couple minutes).

What keycloak version are you on and how are the probes configured?

> Are there any guidelines how to configure the resources?

This is a generally known issue with java workloads on kubernetes.  There's an alpha feature which can help with this 
https://kubernetes.io/blog/2023/05/12/in-place-pod-resize-alpha/#java-processes-initialization-cpu-requirements

Once that is generally supported, you'd ideally want the operator to differentiate between before and after the startup probe has succeeded and reduce the resources as appropriate.  The trade-off being that a keycloak restart could then need to trigger eviction to succeed.

> And are there any things that could be done to improve the startup time (e.g. disabling unused features)?

I think there is still interest in producing native keycloak quarkus images.  @vmuzikar any other thoughts here?

**Reply by NilsEngelbach** [reply]
We are currently running version `22.0.1`, but we deploy the service without the keycloak operator.

The Quickstart for Kubernetes also does not include resource limits / requests as orientation:

https://[raw.githubusercontent.com/keycloak/keycloak-quickstarts/latest/kubernetes/keycloak.yaml](https://raw.githubusercontent.com/keycloak/keycloak-quickstarts/latest/kubernetes/keycloak.yaml)

Disabling features would not improve the startup time?

Thanks for your quick answer, the Kubernetes Feature sounds promising!
I can also understand that fast startup times are not really a common use case / priority for keycloak.

**Comment by shawkins** [comment]
> The Quickstart for Kubernetes also does not include resource limits / requests as orientation:

There are a couple of open issue wrt resources for the operator: #20107 and #21996

> Disabling features would not improve the startup time?

I'm not sure by how much.  Looping in @mabartos as he's looked at startup timings recently.

**Comment by mloiseleur** [comment]
It's possible to reduce memory request down to ~ 350 Mi, with java options:

```yaml
         env:
            - name: JAVA_OPTS_APPEND
              value: -server -Xms196m -Xmx256m
```
